File: jambeth/jAmbeth.Merge/test/de/osthus/ambeth/copy/ObjectCopierTest.java
package de.osthus.ambeth.copy;

import java.util.Arrays;
import java.util.Date;

import junit.framework.Assert;

import org.junit.Test;

import de.osthus.ambeth.ioc.ObjectCopierModule;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;
import de.osthus.ambeth.testutil.AbstractIocTest;
import de.osthus.ambeth.testutil.TestModule;
import de.osthus.ambeth.util.ParamChecker;

/**
 * Tests the ObjectCopier-Bean functionality There is only the ObjectCopierModule needed to allow isolated tests
 */
@TestModule(ObjectCopierModule.class)
public class ObjectCopierTest extends AbstractIocTest
{
	@SuppressWarnings("unused")
	@LogInstance
	private ILogger log;

	protected IObjectCopier fixture;

	@Override
	public void afterPropertiesSet() throws Throwable
	{
		super.afterPropertiesSet();

		ParamChecker.assertNotNull(fixture, "fixture");
	}

	/**
	 * Injected fixture to test
	 * 
	 * @param fixture
	 */
	public void setFixture(IObjectCopier fixture)
	{
		this.fixture = fixture;
	}

	@Test
	public void cloneInteger()
	{
		Integer original = new Integer(5);
		Integer clone = fixture.clone(original);
		Assert.assertSame(original, clone);
	}

	@Test
	public void cloneLong()
	{
		Long original = new Long(5);
		Long clone = fixture.clone(original);
		Assert.assertSame(original, clone);
	}

	@Test
	public void cloneDouble()
	{
		Double original = new Double(5);
		Double clone = fixture.clone(original);
		Assert.assertSame(original, clone);
	}

	@Test
	public void cloneFloat()
	{
		Float original = new Float(5);
		Float clone = fixture.clone(original);
		Assert.assertSame(original, clone);
	}

	@Test
	public void cloneByte()
	{
		Byte original = new Byte((byte) 5);
		Byte clone = fixture.clone(original);
		Assert.assertSame(original, clone);
	}

	@Test
	public void cloneCharacter()
	{
		Character original = new Character((char) 5);
		Character clone = fixture.clone(original);
		Assert.assertSame(original, clone);
	}

	@Test
	public void cloneBoolean()
	{
		Boolean original = new Boolean(true);
		Boolean clone = fixture.clone(original);
		Assert.assertSame(original, clone);
	}

	@Test
	public void cloneDate()
	{
		Date original = new Date(System.currentTimeMillis() - 1000);
		Date clone = fixture.clone(original);
		Assert.assertNotSame(original, clone);
		Assert.assertEquals(original, clone);
	}

	@Test
	public void cloneMaterial()
	{
		StringBuilder original = new StringBuilder("abc");
		StringBuilder clone = fixture.clone(original);
		Assert.assertNotSame(original, clone);
		Assert.assertEquals(original.toString(), clone.toString());
	}

	@Test
	public void cloneByteArrayNative()
	{
		byte[] original = new byte[] { 5, 4, 3, 2, 1 };
		byte[] clone = fixture.clone(original);
		Assert.assertNotSame(original, clone);
		Arrays.equals(original, clone);
	}

	@Test
	public void cloneByteArray()
	{
		Byte[] original = new Byte[] { 5, 4, 3, 2, 1 };
		Byte[] clone = fixture.clone(original);
		Assert.assertNotSame(original, clone);
		Arrays.deepEquals(original, clone);
	}

	@Test
	public void cloneArrayOfArrays()
	{
		Object[][] original = new Object[][] { new Object[] { new Integer(5) }, new Object[] { new Long(6), new Double(7) } };
		Object[][] clone = fixture.clone(original);
		Assert.assertNotSame(original, clone);
		Arrays.deepEquals(original, clone);
	}
}

File: ambeth/Ambeth.Merge.Test/ambeth/copy/ObjectCopierTest.cs
using System;
using System.Text;
using De.Osthus.Ambeth.Copy;
using De.Osthus.Ambeth.Log;
using De.Osthus.Ambeth.Testutil;
using De.Osthus.Ambeth.Util;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace De.Osthus.Ambeth.Ioc
{
    [TestClass]
    [TestModule(typeof(ObjectCopierModule))]
    public class ObjectCopierTest : AbstractIocTest
    {
        [LogInstance]
        public new ILogger Log { private get; set; }

        public IObjectCopier ObjectCopier { protected get; set; }

        public override void AfterPropertiesSet()
        {
            base.AfterPropertiesSet();

            ParamChecker.AssertNotNull(ObjectCopier, "ObjectCopier");
        }

        [TestMethod]
        public void cloneInteger()
        {
            InitManually(GetType());
            int? original = 5;
            int? clone = ObjectCopier.Clone(original);
            Assert.AssertEquals(original, clone);
        }

        [TestMethod]
        public void cloneLong()
        {
            InitManually(GetType());
            long? original = 5;
            long? clone = ObjectCopier.Clone(original);
            Assert.AssertEquals(original, clone);
        }

        [TestMethod]
        public void cloneDouble()
        {
            InitManually(GetType());
            Double original = 5;
            Double clone = ObjectCopier.Clone(original);
            Assert.AssertEquals(original, clone);
        }

        [TestMethod]
        public void cloneFloat()
        {
            InitManually(GetType());
            Single original = 5;
            Single clone = ObjectCopier.Clone(original);
            Assert.AssertEquals(original, clone);
        }

        [TestMethod]
        public void cloneByte()
        {
            InitManually(GetType());
            byte? original = 5;
            byte? clone = ObjectCopier.Clone(original);
            Assert.AssertEquals(original, clone);
        }

        [TestMethod]
        public void cloneCharacter()
        {
            InitManually(GetType());
            char? original = (char)5;
            char? clone = ObjectCopier.Clone(original);
            Assert.AssertEquals(original, clone);
        }

        [TestMethod]
        public void cloneBoolean()
        {
            InitManually(GetType());
            bool original = true;
            bool clone = ObjectCopier.Clone(original);
            Assert.AssertEquals(original, clone);
        }

        [TestMethod]
        public void cloneDate()
        {
            InitManually(GetType());
            DateTime original = new DateTime(DateTime.Now.Ticks - 1000);
            DateTime clone = ObjectCopier.Clone(original);
            Assert.AssertNotSame(original, clone);
            Assert.AssertEquals(original, clone);
        }

        [TestMethod]
        public void cloneMaterial()
        {
            InitManually(GetType());
            StringBuilder original = new StringBuilder("abc");
            StringBuilder clone = ObjectCopier.Clone(original);
            Assert.AssertNotSame(original, clone);
            Assert.AssertEquals(original.ToString(), clone.ToString());
        }

        [TestMethod]
        public void cloneByteArrayNative()
        {
            InitManually(GetType());
            byte[] original = new byte[] { 5, 4, 3, 2, 1 };
            byte[] clone = ObjectCopier.Clone(original);
            Assert.AssertNotSame(original, clone);
            CollectionAssert.AreEqual(original, clone);
        }

        [TestMethod]
        public void cloneByteArray()
        {
            InitManually(GetType());
            Byte[] original = new Byte[] { 5, 4, 3, 2, 1 };
            Byte[] clone = ObjectCopier.Clone(original);
            Assert.AssertNotSame(original, clone);
            CollectionAssert.AreEqual(original, clone);
        }

        [TestMethod]
        public void cloneArrayOfArrays()
        {
            InitManually(GetType());
            Object[][] original = new Object[][] { new Object[] { (int)5 }, new Object[] { (long)6, (Double)7 } };
            Object[][] clone = ObjectCopier.Clone(original);
            Assert.AssertNotSame(original, clone);
            for (int a = original.Length; a-- > 0; )
            {
                CollectionAssert.AreEqual(original[a], clone[a]);
            }
        }
    }
}